---
title: "Untitled"
output: html_document
---

```{r}
library(params)
require(ngsflows)

wd='/rsrch1/iacs/tmp/illumina_platinum/50x/NA12877/splt'
out = create_fq_sheet(x=wd, format="{{samplename}}_{{read}}.{{num}}.fastq", ext = "fastq")
params::kable(head(out))

```


## create commands for alignment preproc and haplo
```{r}

require(ngsflows)
wd = "/rsrch2/iacs/ngs_runs/150806_SN1120_0348_BC79KEACXX/fastqs/Y76I8n1Y76/Project_Pancreatic-MS132/Sample_GLizee-Pancreatic-MS132-MP013normalDNA"
out = create_fq_sheet(x=wd)
#params::kable(head(out))

load_opts("~/Dropbox/public/github_ngsflows/inst/conf/ngsflows_mda.conf", 
          check = FALSE)
source("~/Dropbox/public/github_ngsflows/inst/pipelines/fastq_bam_bwa.R")
#source(fetch_pipes('fastq_bam_bwa')$pip)

fqs1 = head(subset(out, read == 1)$file, 2)
fqs2 = head(subset(out, read == 2)$file, 2)

set_opts(samplename = "MS132")

#debug(fastq_bam_bwa)
f_merge = fastq_bam_bwa(fqs1, fqs2, outfile = out$samplename[1])
f_preproc = bam_preprocess(f_merge$outfile)
f_hapl = haplotyper(f_preproc$outfile)

flowmat = rbind(f_merge$flowmat, f_preproc$flowmat, f_hapl$flowmat)

flowdef = as.flowdef("~/Dropbox/public/github_ngsflows/inst/pipelines/fastq_haplotyper.def")
fobj = to_flow(flowmat, flowdef, flowname = "fastq_haplotyper", flow_run_path = "flowr_test")
submit_flow(fobj, execute = TRUE)


## publication
plot_flow(flowdef, detailed = FALSE, width = 12, height = 10, pdffile = "~/Dropbox/projects/papers_flow/paper_images/fastq_haplotyper.pdf")

plot_flow(flowdef, pdffile = "~/Dropbox/public/github_ngsflows/inst/pipelines/fastq_haplotyper.pdf")

View(flowmat[c(1,2, 16, 17, 33, 49, 66:71), ])

if(FALSE){
  flowdef = to_flowdef(flowmat)
  write_sheet(flowdef, "~/Dropbox/public/github_ngsflows/inst/pipelines/fastq_haplotyper.def")
}

```
